### ISSUE TITLE
Anonymous states

### ISSUE THREAD

**Issue by emilyyyylime** [typst/typst#2425]
### Description

I wrote a whole thing and it got deleted because I forgot to submit the issue and I can't bother redoing it but I would hope for something like
```typc
let x = state(initial-value) // No way to access this state besides through `x`
let y = state(initial-value, key: "some-y-state-key") // Any state constructed with the same key will be associated with `y`.
```

This will break existing code, but using a named argument for the key makes existing code a hard error always, rather than things just not working sometimes (specifically with string-valued states)
There shouldn't be a way to construct a state with no arguments. If no initial value is provided, the state ought to be referencing some existing state in the document.

If you can think of a backwards compatible way to make the state optional I'd be open to hear it

### Use Case

To avoid name collisions, and because usually there's no use for multiple states with the same key, so it just imposes a quirky and potentially brittle api

**Comment by PgBiel** [comment]
Due to the way `state` works, I believe there's no way around having to use _some_ key, but there could certainly be some sort of "auto-generated" internal key which can never collide with custom user keys, perhaps with some sort of hash.

**Comment by laurmaedje** [comment]
Due to function purity constraints, there are some limits to what we can do here. However, it would be possible to bind the key to the state's call site (more specifically, it's span). Then, two different state calls will always have unique keys, but one state call in a loop will always have the same key unless disambiguated by the user. Then, `state` itself isn't technically pure anymore, but a user function that calls `state` internally remains pure.

**Comment by emilyyyylime** [comment]
Yeah that's definitely something I had imagined, it sounds fair to me

**Comment by laurmaedje** [comment]
The `step` function itself is pure, both invocations of it return the same value. You could store the return value of `step()` in a variable and use it in place of the two individual `step()` calls and the result would be the same. What's impure is the runtime that renders the content resulting from your code. This is similar to IO Actions in Haskell.

**Comment by emilyyyylime** [comment]
Additionally, you may think of display as taking a hidden location parameter and using it to determine the state of the state at that point in the document

also, this isn't really related to my issue, is it?

**Comment by timjs** [comment]
@emilyyyylime: already moved this to discussion #2501 to not pollute your issue 🙂

**Comment by dccsillag** [comment]
Regarding adding span as an implicit argument to `state`, I think that's quite dangerous; for example, consider:

```typst
#let create_state(i) = state(i + 2)

#let state1 = create_state(5)
#let state2 = create_state(9)
```

I'd expect `state1` and `state2` to be different states, but if they are implicitly disambiguated by their span then they will actually be the same state.

**Comment by laurmaedje** [comment]
That's the question here: is it better or worse than the current way? Ultimately, because of function purity it's just hard to magically generate a unique state.

It could perhaps be made more explicit though. With a keyword or function that gives an id based on span which can be passed to state().

**Comment by astrale-sharp** [comment]
> Regarding adding span as an implicit argument to `state`, I think that's quite dangerous; for example, consider:
> 
> ```typst
> #let create_state(i) = state(i + 2)
> 
> #let state1 = create_state(5)
> #let state2 = create_state(9)
> ```
> 
> I'd expect `state1` and `state2` to be different states, but if they are implicitly disambiguated by their span then they will actually be the same state.

Here create_state creates a new state, hence we would like to use this call for the span, but can we actually know which call/reference to create state or not?

**Comment by dccsillag** [comment]
Well, I'm drawing a parallel to what code that goes something like this would do in a "typical" language:

```javascript
function create_state(i) {
    return new State(i + 2);
}

let state1 = create_state(5)
let state2 = create_state(9)
```

So I don't think it's of much use to tie the state creation to some particular function call -- it's a bit more like there is some hidden variables being implicitly thread through the program, which get updated upon each state creation.

**Comment by astrale-sharp** [comment]
But then why is the span the same if it comes from 2 different calls?

**Comment by dccsillag** [comment]
Because the span is the actual location of the `state` call in the source code. Because the actual `state` call is in the same position in the source code (line 1, column 24), they would end up becoming the same state. (This is what I'm critiziing, and really believe should not be the case.)

**Comment by astrale-sharp** [comment]
Thing is span is instrumental to memoization at the moment and I don't know if it's easy (probs very very not) to remove it, so what I'm suggesting instead is that we change the used span to where we perceive the state is being created to have a small change still work in that scenario

**Comment by dccsillag** [comment]
Right. I just have no idea how we could figure that out. It may not even be possible in general.

**Comment by laurmaedje** [comment]
> Thing is span is instrumental to memoization at the moment

@astrale-sharp What exactly do you mean by this?

> I'm suggesting instead is that we change the used span to where we perceive the state is being created to have a small change still work in that scenario

> I just have no idea how we could figure that out. It may not even be possible in general.

I think no matter whether we use the span of `state` itself or any of its callees up the stack, there can be collisions where the full call stack is identical. A simple example is `(1, 2).map(state)` where there isn't even another call involved.

**Comment by astrale-sharp** [comment]
> @astrale-sharp What exactly do you mean by this?

My current mental model is that when a typst function is called, comemo sees (unique span, arguments) and if they both are cached already, it gives the result.


> 
> I think no matter whether we use the span of `state` itself or any of its callees up the stack, there can be collisions where the full call stack is identical. A simple example is `(1, 2).map(state)` where there isn't even another call involved.

:thinking:

**Comment by laurmaedje** [comment]
> My current mental model is that when a typst function is called, comemo sees (unique span, arguments) and if they both are cached already, it gives the result.

Comemo generally has two kinds of arguments: Tracked ones (e.g. the world, not really relevant here), and hashed ones. If the tracked arguments are equivalent and the hashed ones have the same hash as in a previous call, comemo yields that call's result. Comemo itself does not care about spans. The spans are designed to be stable through edits to not hurt memoization too badly, but comemo would if anything work better without any spans. They are a necessary "evil" for useful diagnostics.

**Comment by astrale-sharp** [comment]
I see, does comemo has to be aware of span then? or can we just not avoid it cause they're here?

**Comment by laurmaedje** [comment]
Comemo isn't aware of spans in particular, they are just part of the hashed data. They need to be part of that or else the memoization results would be invalid.

**Comment by astrale-sharp** [comment]
Thanks for the explanation!

**Comment by Kannen** [comment]
State keys belong to a global namespace that is shared will all package and files that are imported or included. That is not a scalable solution. Worst than that, 2 calls to the `state` function can be done with the same `key`, causing interference. That is as bad as a mutable static in rust. This causes race condition.

It already happen in a document mine, and I took me hours to find out what was happening. Two states where declared with the same key and used concurrently. The result was random state. no warning, no errors...

Nowaday, I generate unique keys this way:
```
#let auto-key-counter = counter("auto-key-counter")
#let make-key() = {
  repr(context {
  auto-key-counter.step()
  let c = auto-key-counter.display()
  "auto-key-" + str(c)
  })
}
```

So, if it is possible to do it in typst code, why not implementing it? It would be cleaner no?

**Comment by emilyyyylime** [comment]
Well, speaking of namespace collisions, one simple prototype solution could simply be to provide namespacing based on which package a state is created in. That'd already bring us quite far.

**Comment by laurmaedje** [comment]
That goes into a similar direction as my plan, which would be to integrate the exact source code location into the key.

**Comment by Kannen** [comment]
@laurmaedje, Sorry, I am wrong, it does not work.

**Comment by MDLC01** [comment]
I did not read the entire thread, but I don't think this was considered: using a combination of the span and the value of all variables in scope as the key if no string is provided.

**Comment by T0mstone** [comment]
I think the way I did it in #6697 was also not mentioned yet. I can't think of a good short summary, so just read the PR description for what I came up with.

**Comment by laurmaedje** [comment]
> I did not read the entire thread, but I don't think this was considered: using a combination of the span and the value of all variables in scope as the key if no string is provided.

I think that would make it even more strangely magic without fully eliminating the possibility of collisions.

From another perspective though, I'm not sure it's even desirable to find the most unique key: As a user that expects Typst functions to be pure, that might be quite surprising. 

At the very least, we must ensure that a state call wrapped in a function does not make the function impure.

**Comment by laurmaedje** [comment]
I also had this thought of having a keyword that must be passed to `state` so that `state` itself remains pure, but that's probably overkill, also considering that stuff like `read` is already similarly locally impure with path resolution. (Locally impure here meaning that wrapper functions around `read` remain pure.)

**Comment by bluss** [comment]
I have a use case that I would like to make easier to achieve using anonymous states. It's this thing where you feel compelled to use a counter in the state name (which is not a recommended practice - rather the opposite). Could anonymous states help us get rid of the counter, making the typst code better and more performant? (Or if I'm mistaken - is there an easier way to do this?)

You'd want to get a new anonymous state for every time the function is called, in this case.

```typ
#let lrc = counter("lraverage")
// This function finds all `math.lr` in the body and ensures they have the same size (the average or max size, depending on method)
#let lraverage(body, method: "average", scale: 100%) = {
  lrc.step()
  context {
    let st = state("lraverage:" + str(lrc.get().first()))
    st.update(())
    show math.lr: it => {
      if (it.size != 100%) { return it }
      context {
        let sz = measure(math.equation(it.body, block: true))
        st.update(d => (..d, sz.height))
        let sizes = st.final()
        if sizes.len() > 0 {
          let size = if method == "average" {
            sizes.sum()/sizes.len()
          } else if method == "max" {
            calc.max(..sizes)
          }
          math.lr(it.body, size: size * scale)
        } else {
          it
        }
      }
    }
    body
  }
}

#lraverage[
$ X & = (1 + a/2)(1 + b/3) $
]
```

<img width="300" height="172" alt="screenshot of above equation" src="https://github.com/user-attachments/assets/f5a6d793-65f3-41a5-a137-27404f5fc0ad" />

**Comment by laurmaedje** [comment]
@bluss One way to do this without generated state keys is by placing the height in labelled metadata elements and then querying the metadata elements between the start and end of your body (using before/after).

**Comment by bluss** [comment]
Ok. When I use metadata and query, what I see is that I need to place an end marker, then query for the location of the end marker before I can use the end marker in a another `selector.before(endmarker)` - that's the kind of query that depends on a query that I'd be wary to use. Maybe there's another way, or is that good, or is that something that could be improved in the future?

(*) I'd clarify that I could use `before(<endmarker>)` as a label. But to be re-usable it needs to be a unique label, so then we need a counter in the label, maybe that works?

**Comment by laurmaedje** [comment]
A query that includes the result of another query as part of the selector isn't necessarily a problem as long as both can be resolved in the same iteration. Something like `let end = query(selector(<endmarker>).after(here())).first().location()` and then `query(selector(metadata).after(here()).before(end)` _should_ be fine.

**Comment by bluss** [comment]
Thank you. Since this was now off topic, I'm sorry for the noise, but I'll add that I then found this nice pattern then with all this taken together:

```typ
#let all-rules = query(selector.or(setrowlab, setcollab)
  .after(here()).before(selector(endlab).after(here())))
```

- `setrowlab, setcollab`  are labels of what we are looking for
- and `selector(endlab).after(here())` is a valid `before` spec due to the composability of selectors and similar things, and it seems to do the right thing when there are multiple `endlab`, except if begin/ends are nested, then you need a numbered label.

**Comment by patrick-kidger** [comment]
For those who found this issue like I did: at least for `counter`s (but not `state`s), then I have found a way to generate unique `counter`s without needing unique strings.

These are available in [typsy](https://github.com/patrick-kidger/typsy/), and usable like so:
```typst
#import "@preview/typsy:0.1.0": safe-counter

#let my-counter1 = safe-counter(()=>{})
#let my-counter2 = safe-counter(()=>{})
```
As you can tell, this uses a bit of a trick – each anonymous function has a unique identity, and we can launder that into a unique counter!

**Comment by patrick-kidger** [comment]
Follow-up: I figured out how to do anonymous `state`s as well! As of typsy:0.2.0:

```typst
#import "@preview/typsy:0.2.0": safe-state

#let my-state1 = safe-state(()=>{}, "some default value")
#let my-state2 = safe-state(()=>{}, "another default value")
```

(Note that these use the usual dictionary trick for emulating methods, so getting the value is done by e.g. `(my-state.get)()`, note the extra pair of brackets.)

**Comment by Lege19** [comment]
I've taken to using a uuid snippet in my editor to just append a uuid to the end of the descriptive part of the name of labels, states, and counters. It creates a bit of noise but it makes the behavior extremely clear.

I genuinely think this is the best solution.

I think there's not much point in supporting something tied to spans because that's effectively just this but trading a bit of noise for a lot of confusion in how it behaves around functions.
